pkg/chunk/storage_client.go

```diff
@@ -30,6 +31,9 @@ type ObjectAndIndexClient interface {
 // WriteBatch represents a batch of writes.
 type WriteBatch interface {
 	Add(tableName, hashValue string, rangeValue []byte, value []byte)
+	Increment(tableName, hashValue string, rangeValue []byte, value int64)
```

Comment:
I think you should consider creating a separate client for deletions instead of using the index client.

First off it will avoid having a bunch of clients with deletes unimplemented. Also using a dedicated client will allow for a cleaner interface to be designed. Also it will make it easier to design something that can be used more easily with a the internal KVStore which could be a good fit for this use case.

Reply:
KVStore was an option but I didn't want to add an additional dependency so I went with reusing existing code. Please let me know what do you think about it.

Reply:
What do you mean by an additional dependency? Couldn't this be implemented on top of a ETCD/Consul/Gossip kvstore, which already exist?

Reply:
Maybe gen numbers can be stored but we also have to store delete plans and delete requests which would be too much for gossip at least. Also I wouldn't consider them as reliable as index stores for storing data which won't get lost.

Reply:
Cache Gen numbers should be stored in a KVStore. We can take advantage of the `CAS` interface. We can also do it in a way where the read/write path should know nothing about the cache gen number outside of the cache package. All that would be needed is another layer of cache with a ETCD/Consul KVStore embedded in it. This same KVStore would also be used in the Purger.